arch/arm/stm32: Basic STM32F7 family support

pic16f887 opened the pull request:
Related to issue #6981.
Added support for STM32F746XG subfamily.

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
arch/arm/stm32: Basic STM32F7 support
Related to issue #6981

erwango commented on arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,42 @@
+config SOC_SERIES
+	default stm32f7
+
+if USB
```
Remove USB as it is not yet tested/implemented on this series (limited to F4 for now)

erwango commented on arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,42 @@
+
+endif #USB
+
+if I2C && (I2C_1 || I2C_2 || I2C_3 || I2C_4)
```
Remove I2C, not yet tested/supported on F7 series

erwango commented on arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.stm32f746xg:
```diff
@@ -0,0 +1,32 @@
+
+endif # GPIO_STM32
+
+if ENTROPY_GENERATOR
```
Remove for now. You could add it in a follow up PR.

erwango commented on arch/arm/soc/st_stm32/stm32f7/dts.fixup:
```diff
@@ -0,0 +1,102 @@
+#define CONFIG_UART_STM32_PORT_6_IRQ_PRI		ST_STM32_USART_40011400_IRQ_0_PRIORITY
+#define PORT_6_IRQ								ST_STM32_USART_40011400_IRQ_0
+
+#define CONFIG_I2C_1_BASE_ADDRESS				ST_STM32_I2C_V1_40005400_BASE_ADDRESS
```
Remove I2C, SPI, USB and flash

erwango commented on arch/arm/soc/st_stm32/stm32f7/flash_registers.h:
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2018 Yurii Hamann
+ * Copyright (c) 2016 Linaro Limited.
```
Keep only your copyright, apply on all files

erwango commented on arch/arm/soc/st_stm32/stm32f7/soc.h:
```diff
@@ -0,0 +1,62 @@
+#include <stm32f7xx_ll_usart.h>
+#endif
+
+#ifdef CONFIG_I2C
```
Remore I2C

erwango commented on arch/arm/soc/st_stm32/stm32f7/soc.h:
```diff
@@ -0,0 +1,62 @@
+#include <stm32f7xx_ll_i2c.h>
+#endif
+
+#ifdef CONFIG_ENTROPY_STM32_RNG
```
Remove Entropy

erwango commented on arch/arm/soc/st_stm32/stm32f7/soc.h:
```diff
@@ -0,0 +1,62 @@
+#include <stm32f7xx_ll_rng.h>
+#endif
+
+#ifdef CONFIG_IWDG_STM32
```
Remove

erwango commented on arch/arm/soc/st_stm32/stm32f7/soc.c:
```diff
@@ -0,0 +1,64 @@
+ *
+ * @return 0
+ */
+static int st_STM32F7_init(struct device *arg)
```
keep lower case

erwango commented on drivers/pinmux/stm32/pinmux_stm32f7.h:
```diff
@@ -0,0 +1,308 @@
+ * @file Header for STM32F7 pin multiplexing helper
+ */
+
+/* Port A */
```
Clean everything which is not UART

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -295,7 +295,7 @@ static int uart_stm32_init(struct device *dev)
 	LL_USART_Enable(UartInstance);
 
-#if !defined(CONFIG_SOC_SERIES_STM32F4X) && !defined(CONFIG_SOC_SERIES_STM32F1X)
+#if !defined(CONFIG_SOC_SERIES_STM32F4X) && !defined(CONFIG_SOC_SERIES_STM32F7X) && !defined(CONFIG_SOC_SERIES_STM32F1X)
```
Take care to the 80 char limit

erwango commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,262 @@
+			label= "I2C_1";
+		};
+
+		i2c2: i2c@40005800 {
```
clean up i2c spi and USB

erwango reviewed (changes requested):
Hi @pic16f887 ,

Thanks for this work.
Some initial comments on the pull request:
 * focus your work on uart, gpio, exti and clock-control drivers which are required for basic testing (hello_world and blinky). Clean up everything is relates to USB, I2C, SPI, .... It will ease to focus and fasten the review process.
  *Split your work into coherent commit (arch, dts, drivers/clock_control, drivers/uart, ...)
  * Last: please port a F7 board (in a dedicated commit) so the code could at least be compiled

Also, you should add your Signed-Off in the commit message

Thanks again!

ydamigos commented on arch/arm/soc/st_stm32/stm32f7/Kconfig.series:
```diff
@@ -0,0 +1,20 @@
+
+config SOC_SERIES_STM32F7X
+	bool "STM32F7x Series MCU"
+	select CPU_CORTEX_M
```
You should remove this. CPU_CORTEX_M is selected when you select CPU_CORTEX_M7.

Commit:
Revert "arch/arm/stm32: Basic STM32F7 support"

This reverts commit 0b1522e3d31a2c65a2480ef67d09aa94754d33af.

Commit:
arch: arm: stm32: Basic STM32F7 family support
Includes support for STM32F746xG subfamily
Related to the issue #6981

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
arch: arm: stm32: Basic STM32F7 family support
Includes support for STM32F746xG subfamily
Related to the issue #6981

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
boards: arm: stm32: Support for STM32F746G Discovery board
Related to issue #6982

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: clock_control: STM32F7 family clock control support

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: clock_control: STM32F7 family clock control support

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: gpio: GPIO support for STM32F7 family

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: gpio: GPIO support for STM32F7 family

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: interrupt_controller: STM32F7 interrupt controller support

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: interrupt_controller: STM32F7 interrupt controller support

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: pinmux: STM32F7 pinmux support

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: pinmux: STM32F7 pinmux support

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: serial: STM32F7 UART support

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: serial: STM32F7 UART support

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
dts: arm: st: STM32F7 family device tree

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
dts: arm: st: STM32F7 family device tree

Signed-off-by: Yurii Hamann <yurii@hamann.site>

pic16f887 commented:
Hello @erwango ,
I have split the commit into smaller ones as you asked me. I added support for STM32F746G Discovery board. I tested it on on "hello_world" and "blinky" demo.

dbkinder commented on boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst:
```diff
@@ -0,0 +1,200 @@
+Overview
+********
+
+The discovery kit enables a wide diversity of applications taking benefit from audio, multi-sensor support, graphics, security, video and high-speed connectivity features.
```
Preference is for lines to be <80 characters

Change the ending to read:

   ... security, video, and high-speed connectivity features.  Important board features include:

dbkinder commented on boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst:
```diff
@@ -0,0 +1,200 @@
+System Clock
+============
+
+STM32F746G System Clock could be driven by internal or external oscillator,
```
lots of missing articles (the, a, an):

    The STM32F746G System Clock can be driven by an internal or external oscillator,
    as well as by the main PLL clock. By default, the System clock is driven by the PLL
    clock at 216MHz, driven by a 25MHz high speed external clock.

dbkinder commented on boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst:
```diff
@@ -0,0 +1,200 @@
+Serial Port
+===========
+
+The STM32F746G Discovery kit has up to 8 UARTs. The Zephyr console output is assigned to UART1.
```
I suspect the kit has a fixed number of UARTs, so it's better to say:

    The STM32F746G Discovery kit has 8 UARTs.

pic16f887 replied:
@dbkinder, 
I am not sure that all 8 UARTs are available. Many ports are occupied by LCD display with touchscreen and other peripherals.

dbkinder commented on boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst:
```diff
@@ -0,0 +1,200 @@
+===========
+
+The STM32F746G Discovery kit has up to 8 UARTs. The Zephyr console output is assigned to UART1.
+Default settings are 115200 8N1.
```
``Default communication settings...``

dbkinder commented on boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst:
```diff
@@ -0,0 +1,200 @@
+========
+
+STM32F746G Discovery kit includes an ST-LINK/V2 embedded debug tool interface.
+This interface is supported by the openocd version included in Zephyr SDK.
```
``in the Zephyr``

dbkinder reviewed (changes requested):
Thanks for the documentation.  A few fixes...

Commit:
Revert "boards: arm: stm32: Support for STM32F746G Discovery board"

This reverts commit b249ec0da69ca7cd37aed628e45aced542f37352.

ydamigos commented on arch/arm/soc/st_stm32/stm32f7/soc.h:
```diff
@@ -0,0 +1,50 @@
+ * @file SoC configuration macros for the ST STM32F7 family processors.
+ *
+ * Based on reference manual:
+ *   RM0368 Reference manual STM32F401xB/C and STM32F401xD/E
```
Please add STM32F7 reference manuals.

ydamigos commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,140 @@
+#include <arm/armv7-m.dtsi>
+#include <st/mem.h>
+#include <dt-bindings/clock/stm32_clock.h>
+#include <dt-bindings/i2c/i2c.h>
```
You could remove this.

ydamigos commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,140 @@
+	};
+
+	soc {
+		flash-controller@40023c00 {
```
Do we support flash controller for stm32f7 family? I couldn't find it in the flash drivers. If not, please remove this node.

ydamigos reviewed (changes requested):
You should cleanup your PR. Please drop the commits that you revert and fix your commits body to make shippable happy.

dbkinder reviewed (changes requested):
documentation seems to be missing now?  And there are problems with the commit messages reported by shippable: (https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/14066/5/tests)

Commit:
boards: arm: stm32: Basic support for STM32F746G Discovery board
The patch contains basic code support, documentation, dts file
and openocd configuration for STM32F746G discovery board
Patch related to issue #6982

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
boards: arm: stm32: Basic support for STM32F746G Discovery board
The patch contains basic code support, documentation, dts file
and openocd configuration for STM32F746G discovery board
Patch related to issue #6982

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
Merge branch 'master' of github.com:pic16f887/zephyr
